File: src/DSA/Searching/SearchingProblems.cs
using System;

namespace DSA.Searching
{
    public static class SearchingProblems
    {
        // Binary Search Algorithm Iterative version
        public static int BinarySearchIterative(int[] arr, int x)
        {
            int n = arr.Length;
            int l = 0, h = n - 1;

            while (l <= h)
            {
                int mid = l + (h - l) / 2;
                if (arr[mid] == x)
                    return mid;

                if (x < arr[mid])
                {
                    h = mid - 1;
                }
                else
                {
                    l = mid + 1;
                }
            }
            return -1;
        }

        // Binary Search Algorithm Recursive version
        public static int BinarySearchRecursive(int[] arr, int l, int h, int x)
        {
            if (l > h)
            {
                return -1;
            }

            int mid = l + (h - l) / 2;
            if (arr[mid] == x)
                return mid;

            if (x < arr[mid])
            {
                return BinarySearchRecursive(arr, l, mid - 1, x);
            }
            else
            {
                return BinarySearchRecursive(arr, mid + 1, h, x);
            }
        }

        // Find first occurence or left most index of an element in a sorted array
        public static int FirstOccurence(int[] arr, int l, int h, int x)
        {
            if (l > h)
            {
                return -1;
            }

            int mid = l + (h - l) / 2;
            if (arr[mid] == x && (mid == 0 || arr[mid - 1] != x))
            {
                return mid;
            }

            if (x <= arr[mid])
            {
                return FirstOccurence(arr, l, mid - 1, x);
            }
            else
            {
                return FirstOccurence(arr, mid + 1, h, x);
            }
        }

        // Find last occurence or right most index of an element in a sorted array
        public static int LastOccurence(int[] arr, int l, int h, int x)
        {
            if (l > h)
            {
                return -1;
            }

            int mid = l + (h - l) / 2;
            int n = arr.Length;
            if (arr[mid] == x && (mid == n - 1 || arr[mid + 1] != x))
            {
                return mid;
            }

            if (x >= arr[mid])
            {
                return LastOccurence(arr, mid + 1, h, x);
            }
            else
            {
                return LastOccurence(arr, l, mid - 1, x);
            }
        }

        // Find no. of occurences of an element in a sorted array
        public static int CountOccurences(int[] arr, int x)
        {
            int n = arr.Length;

            int first = FirstOccurence(arr, 0, n - 1, x);
            if(first == -1)
            {
                return 0;
            }

            int last = LastOccurence(arr, 0, n - 1, x);

            return last - first + 1;
        }

        // Find no. of one's in a binary sorted increasing array
        public static int CountOnesInBinarySortedArray(int[] arr, int x)
        {
            int n = arr.Length;

            int first = FirstOccurence(arr, 0, n - 1, 1);

            return n - first + 1;
        }

        //Given a binary sorted non-increasing array arr of size N.You need to print the count of 1's in the binary array.
        public static int CountOnesInSortedDecreasingArray(int[] arr, int n)
        {

            int low = 0;
            int high = n - 1;

            while (low <= high)
            {
                int mid = low + ((high - low) / 2);

                if (arr[mid] == 1 && (mid == n - 1 || arr[mid + 1] != 1))
                {
                    return mid + 1;
                }

                if (arr[mid] == 0)
                {
                    high = mid - 1;
                }
                else
                {
                    low = mid + 1;
                }
            }

            return 0;

        }

        // Find an element in an infinite size sorted array
        public static int FindInInfiniteSizedSortedArr(int[] arr, int x)
        {
            if (arr[0] == x)
            {
                return 0;
            }
            int i = 1;
            while (arr[i] < x)
            {
                i = i * 2;
            }

            if (arr[i] == x)
            {
                return i;
            }

            return BinarySearchRecursive(arr, i / 2 + 1, i - 1, x);
        }

        // Find an element in a sorted rotated array
        public static int FindInSortedRotatedArr_approach1(int[] arr, int x)
        {
            int n = arr.Length;
            int low = 0, high = n - 1;
            while (low <= high)
            {
                int mid = low + (high - low) / 2;

                if (arr[mid] == x)
                {
                    return mid;
                }

                if (arr[mid] > arr[low]) // left half is sorted
                {
                    if (x >= arr[low] && x < arr[mid])
                    {
                        high = mid - 1;
                    }
                    else
                    {
                        low = mid + 1;
                    }
                }
                else // right half is sorted
                {
                    if (x > arr[mid] && x <= arr[high])
                    {
                        low = mid + 1;
                    }
                    else
                    {
                        high = mid - 1;
                    }
                }
            }

            return -1;
        }

        // Find pivot in sorted rotated array
        public static int FindPivot(int[] arr, int low, int high)
        {
            if (low > high) return -1;

            if (low == high) return low;

            int mid = low + (high - low) / 2;
            //check if middle element is pivot

            if (mid < high && arr[mid] > arr[mid + 1])
            {
                return mid;
            }

            if (mid > low && arr[mid] < arr[mid - 1])
            {
                return (mid - 1);
            }

            if (arr[mid] <= arr[low])
            {
                return FindPivot(arr, low, mid - 1);
            }

            return FindPivot(arr, mid + 1, high);
        }

        // Find an element in a sorted rotated array
        public static int PivotedBinarySearch(int[] arr, int x)
        {
            int n = arr.Length;
            int pivot = FindPivot(arr, 0, n - 1);

            if (pivot == -1)
            {
                return BinarySearchRecursive(arr, 0, n - 1, x);
            }

            if (arr[pivot] == x)
            {
                return pivot;
            }

            if (x >= arr[0])
            {
                return BinarySearchRecursive(arr, 0, pivot - 1, x);
            }

            return BinarySearchRecursive(arr, pivot + 1, n - 1, x);
        }

        // Given an integer x. The task is to find the square root of x. If x is not a perfect square, then return floor(√x).
        public static long FindSquaroot(long x)
        {
            if (x == 0 || x == 1) return x;
            long start = 1, end = x, ans = 1;
            while (start <= end)
            {
                long mid = start + (end - start) / 2;
                if (mid * mid == x)
                {
                    return mid;
                }

                if (mid * mid < x)
                {
                    ans = mid;
                    start = mid + 1;
                }
                else
                {
                    end = mid - 1;
                }
            }

            return ans;
        }

        // Find peak element element in a sorted  array
        public static int FindPeakElement(int[] arr, int n, int low, int high)
        {
            int mid = low + (high - low) / 2;
            if ((mid == 0 || arr[mid] >= arr[mid - 1]) && (mid == n - 1 || arr[mid] >= arr[mid + 1]))
            {
                return mid;
            }

            if (mid > 0 && arr[mid - 1] > arr[mid])
            {
                return FindPeakElement(arr, n, low, high - 1);
            }

            return FindPeakElement(arr, n, mid + 1, high);
        }

        //Given an array, check if there exists a pair, whose sum is given sum
        public static bool IsPairWithGivenSum(int[] arr, int start, int end, int x)
        {
            while (start <= end)
            {
                int sum = arr[start] + arr[end];
                if (sum == x)
                {
                    return true;
                }

                if (sum > x)
                {
                    end--;
                }
                else
                {
                    start++;
                }
            }

            return false;
        }

        //Given an array, check if there exists a triplet, whose sum is given sum
        public static bool IsTripletWithGivenSum(int[] arr, int x)
        {
            Array.Sort(arr);

            int n = arr.Length;

            for (int i = 0; i < n - 2; i++)
            {
                if (IsPairWithGivenSum(arr, i + 1, n - 1, x - arr[i]))
                {
                    return true;
                }
            }

            return false;
        }

        // Count pair with a given sum in sorted array
        public static int CountPairWithGivenSum(int[] arr, int start, int end, int givenSum)
        {
            int count = 0;
            while (start < end)
            {
                int currSum = arr[start] + arr[end];
                if (currSum == givenSum)
                {
                    count++;
                }

                if (currSum > givenSum)
                {
                    end--;
                }
                else
                {
                    start++;
                }
            }

            return count;
        }

        // Count triplet with a given sum in an array
        public static int CountTripletsWithGivenSum(int[] arr, int sum)
        {
            Array.Sort(arr);

            int n = arr.Length;
            int totalCount = 0;
            for (int i = 0; i < n; i++)
            {
                int count = CountPairWithGivenSum(arr, i + 1, n - 1, sum - arr[i]);
                totalCount += count;

            }

            return totalCount;
        }

        // check if array has pythagoras triplets
        public static bool CheckForPythagorasTriplet(int[] arr)
        {

            Array.Sort(arr);

            int n = arr.Length;
            for (int i = 0; i < n; i++)
            {
                arr[i] = arr[i] * arr[i];
            }

            for (int i = 0; i < n - 1; i++)
            {
                if (IsPairWithGivenSum(arr, i + 1, n - 1, arr[i]))
                {
                    return true;
                }
            }

            return false; ;
        }

        // Get median of two sorted arrays
        public static double GetMedianOfTwoSortedArrays(int[] arr1, int[] arr2)
        {
            int n1 = arr1.Length;
            int n2 = arr2.Length;

            int begin1 = 0, end1 = n1;
            double res = 0;
            while (begin1 <= end1)
            {
                int i1 = begin1 + ((end1 - begin1) / 2);
                int i2 = n1 + (n2 - n1 + 1) / 2 - i1;

                int min1 = i1 == n1 ? Int32.MaxValue : arr1[i1];
                int max1 = i1 == 0 ? Int32.MinValue : arr1[i1 - 1];

                int min2 = i2 == n2 ? Int32.MaxValue : arr2[i2];
                int max2 = i2 == 0 ? Int32.MinValue : arr2[i2 - 1];

                if (max1 <= min2 && max2 <= min1)
                {
                    if ((n1 + n2) % 2 == 0)
                    {
                        res = (double)(Math.Max(max1, max2) + Math.Min(min1, min2)) / 2;
                    }
                    else
                    {
                        res = (double)Math.Max(max1, max2);
                    }

                    return res;
                }
                else if (max1 > min2)
                {
                    end1 = i1 - 1;
                }
                else
                {
                    begin1 = i1 + 1;
                }
            }

            return res;
        }

        //Given a sorted array arr[] of size N without duplicates, and given a value x. Find the floor of x in given array. Floor of x is defined as the largest element K in arr[] such that K is smaller than or equal to x.
        public static int FindFloor(long[] arr, int left, int right, long x)
        {
            int res = -1;
            while (left <= right)
            {
                int mid = left + (right - left) / 2;

                if (arr[mid] == x)
                {
                    res = mid;
                    break;
                }

                if (arr[mid] > x)
                {
                    right = mid - 1;
                }
                else
                {
                    res = mid;
                    left = mid + 1;
                }
            }
            return res;
        }

        //Minimum Number in a sorted rotated array
        public static int MinInSortedRotatedArr(int[] arr, int low, int high)
        {
            int pivot = FindPivot(arr, low, high);
            int n = high - low + 1;
            return arr[(pivot + 1) % n];
        }

        //You are given an array of N+2 integer elements. All elements of the array are in range 1 to N. Also, all elements occur once except two numbers which occur twice. Find the two repeating numbers.
        public static void TwoRepeated(int[] arr, int n)
        {

            String res = "";
            for (int i = 0; i < n; i++)
            {
                int index = Math.Abs(arr[i]) - 1;
                int val = arr[index];
                if (val < 0)
                    res += " " + arr[i];
                else
                    arr[index] = -arr[index];
            }

            Console.WriteLine(res.Trim());
        }

        // You are given an array of N integers including 0. The task is to find the smallest positive number missing from the array.
        public static int MissingSmallestPositiveNum(int[] arr, int n)
        {
            // seggregate negavtives to left side and positive to right side
            int j = 0;
            for (int i = 0; i < n; i++)
            {
                if (arr[i] <= 0)
                {
                    Util.Swap(arr, i, j);
                    j++;
                }
            }

            int[] arr2 = new int[n - j];
            int k = 0;
            for (int i = j; i < n; i++)
            {
                arr2[k] = arr[i];
                k++;
            }

            // idea: if we treat array element as index's of array and mark those indexes then
            // the first index which is not marked will represent index + 1 the smallest positive missing no. that is missing

            // Mark arr[i] as visited by making 
            // arr[arr[i] - 1] negative. Note that 
            // 1 is subtracted because index start 
            // from 0 and positive numbers start from 1 
            for (int i = 0; i < k; i++)
            {
                int x = Math.Abs(arr2[i]);
                if (x > 0 && x - 1 < k && arr2[x - 1] > 0)
                {
                    arr2[x - 1] = -arr2[x - 1];
                }
            }

            // Return the first index value at which 
            // is positive 
            for (int i = 0; i < k; i++)
            {
                if (arr2[i] > 0)
                {
                    return i + 1;
                }
            }

            return k + 1;
        }

        // You are given heights of consecutive buildings. You can move from the roof of a building to the roof of next adjacent building. 
        //You need to find the maximum number of consecutive steps you can put forward such that you gain an increase in altitude with each step.
        public static int MaxStep(int[] arr, int n)
        {

            if (n == 1)
            {
                return 0;
            }

            int count = 0, max = Int32.MinValue;
            for (int i = 1; i < n; i++)
            {
                if (arr[i - 1] < arr[i])
                {
                    count++;
                }
                else
                {
                    max = Math.Max(max, count);
                    count = 0;
                }
            }

            return Math.Max(max, count);
        }

        // Given an integer array representing the heights of N buildings, the task is to delete N-2 buildings such that the water that can be trapped between the remaining two building is maximum.
        public static int MaxWaterStored(int[] arr, int n)
        {
            int i = 0, j = n - 1;
            int res = Int32.MinValue;
            while (i < j)
            {
                res = Math.Max(res, (j - i - 1) * Math.Min(arr[i], arr[j]));

                if (arr[i] < arr[j])
                    i++;
                else if (arr[j] < arr[i])
                    j--;
                else
                {
                    i++;
                    j--;
                }
            }
            return res;
        }

        //Given an array of positive integers, where elements are consecutive (sorted). 
        //Also, there is a single element which is repeating X (any variable) number of times. 
        //Now, the task is to find the element which is repeated and number of times it is repeated.
        public static void FindRepeatingInConsecutiveSorted(int[] arr, int n)
        {
            int l = 0, h = n - 1;
            int mid = 0;
            while (l < h)
            {
                mid = l + (h - l) / 2;
                if (arr[mid] < (arr[0] + mid))
                {
                    h = mid;
                }
                else
                {
                    l = mid + 1;
                }
            }

            // if no elements were repeated then we can write
            // n = a[n-1] - a[0] + 1
            // but if an element is repeated say x number of times then
            // n - x = a[n-1] - a[0]

            // no of time repeated
            int count = (arr[0] + n - arr[n - 1]);
            Console.WriteLine(arr[l] + " " + count);
        }

        // Given an unsorted array A of size N of non-negative integers, find a continuous sub-array which adds to a given number S.
        public static void SubArraySum(int[] arr, int n, int sum)
        {
            int currSum = arr[0], s = 0, e;
            
            for (e = 1;  e <= n; e++)
            {
                while (currSum > sum && s < e - 1)
                {
                    currSum = currSum - arr[s];
                    s++;
                }
                if(currSum == sum)
                {
                    break;
                }

                if(e<n)
                    currSum = currSum + arr[e];
            }

            if (currSum == sum)
            {
                Console.WriteLine((s + 1) + " " + e);                
            } else {
                Console.WriteLine(-1);
            }
        }
    }
}

File: java/src/Searching/searchingProblems.java
package Searching;

import java.util.Arrays;
import java.util.HashSet;

public class searchingProblems {

    public int BinarySearchUtil(int[] A, int x, int low, int high) {
        // Assume A is sorted
        if (A[0] == x)
            return 0;

        if (A[A.length - 1] == x)
            return A.length - 1;

        int res = -1;
        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (A[mid] == x)
                return mid;

            if (x < A[mid]) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return res;
    }

    public int BinarySearch(int[] A, int x) {
        return BinarySearchUtil(A, x, 0, A.length - 1);
    }

    public int FirstOccurence(int[] A, int x) {
        // Assume A is sorted
        int res = -1;
        int low = 0, high = A.length - 1;
        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (A[mid] == x && (mid == 0 || A[mid] != A[mid - 1]))
                return mid;

            if (x > A[mid]) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return res;
    }

    public int LastOccurence(int[] A, int x) {
        // Assume A is sorted
        int res = -1;
        int low = 0, high = A.length - 1;
        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (A[mid] == x && (mid == A.length - 1 || A[mid] != A[mid + 1]))
                return mid;

            if (x < A[mid]) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return res;
    }

    public int CountOnesInBinarySortedArray(int[] A) {
        int low = 0, high = A.length - 1;

        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (A[mid] == 0) {
                low = mid + 1;
            } else {
                if (mid == 0 || A[mid - 1] == 0) {
                    return A.length - mid;
                }
                high = mid - 1;
            }
        }

        return 0;
    }

    public int SquareRootFloor(int x) {
        int low = 1, high = x;
        int ans = -1;

        while (low <= high) {
            int mid = (low + high) / 2;

            int sq = mid * mid;

            if (sq == x)
                return mid;

            if (sq > x) {
                high = mid - 1;
            } else {
                ans = mid;
                low = mid + 1;
            }
        }

        return ans;
    }

    public int SearchInInfiniteSortedArray(int[] A, int x) {
        // Assume A is sorted and has infinite no. of elements

        if (A[0] == x)
            return 0;

        int i = 1;
        while (A[i] < x) {
            i = i * 2;
        }

        if (A[i] == x)
            return i;

        return BinarySearchUtil(A, x, i / 2 + 1, i - 1);
    }

    // Find Peak Element. i.e Element which is not smaller than it's neighbours
    public int FindPeakElement(int[] A) {
        int low = 0, high = A.length - 1;

        while (low <= high) {
            int mid = (low + high) / 2;

            if ((mid == 0 || A[mid] >= A[mid - 1]) && (mid == A.length - 1 || A[mid] <= A[mid + 1])) {
                return mid;
            }

            if (mid > 0 && A[mid - 1] >= A[mid]) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return -1;
    }

    public void FindPairWithGivenSumInUnsortedArray(int arr[], int sum) {

        HashSet<Integer> s = new HashSet<Integer>();
        for (int i = 0; i < arr.length; ++i) {
            int temp = sum - arr[i];

            if (s.contains(temp)) {
                System.out.println("Pair with given sum " + sum + " is (" + arr[i] + ", " + temp + ")");
            }
            s.add(arr[i]);
        }
    }

    public boolean FindPairWithGivenSumInSortedArray(int A[], int sum) {
        // Asuming A is sorted
        int low = 0, high = A.length - 1;
        while (low < high) {
            if (A[low] + A[high] == sum) {
                return true;
            }

            if (A[low] + A[high] > sum) {
                high--;
            } else {
                low++;
            }
        }

        return false;
    }

    public boolean FindTripletWithGivenSumInSortedArray(int A[], int sum) {
        // Asuming A is sorted
        for (int i = 0; i < A.length; i++) {
            int low = i + 1, high = A.length - 1;
            while (low < high) {
                if (A[low] + A[high] == sum - A[i]) {
                    return true;
                }

                if (A[low] + A[high] > sum - A[i]) {
                    high--;
                } else {
                    low++;
                }
            }
        }

        return false;
    }

    // Median of two sorted arrays
    public double MedianOfTwoSortedArrays(int[] A1, int[] A2) {
        int n1 = A1.length, n2 = A2.length;

        if (n2 < n1) {// Making sure A1 is smaller
            int[] t = A1;
            A1 = A2;
            A2 = t;
        }

        int low = 0, high = n1;

        while (low < high) {
            int i1 = (low + high) / 2;
            int i2 = (n1 + n2 + 1) / 2 - i1;

            int min1 = i1 == n1 ? Integer.MAX_VALUE : A1[i1];
            int max1 = i1 == 0 ? Integer.MIN_VALUE : A1[i1 - 1];

            int min2 = i2 == n2 ? Integer.MAX_VALUE : A2[i2];
            int max2 = i2 == 0 ? Integer.MIN_VALUE : A2[i2 - 1];

            if (max1 <= min2 && max2 <= min1) {
                if ((n1 + n2) % 2 == 0) {
                    return (double) (Math.max(max1, max2) + Math.min(min1, min2)) / 2;

                } else {
                    return (double) Math.max(max1, max2);
                }
            }

            if (max2 > min1) {
                high = i1 - 1;

            } else {
                low = i1 + 1;
            }
        }

        return -1;
    }

    // Median of a Row Wise Sorted Matrix Java
    public int MedianOfRowWiseSortedMatrix(int[][] mat, int r, int c) {
        int min = mat[0][0], max = mat[0][c - 1];

        for (int i = 0; i < r; i++) {
            if (mat[i][0] < min) {
                min = mat[i][0];
            }

            if (mat[i][c - 1] > max) {
                max = mat[i][c - 1];
            }
        }

        int medianPos = (r * c + 1) / 2;
        while (min < max) {
            int mid = (min + max) / 2;

            int midPos = 0;

            for (int i = 0; i < r; i++) {
                int pos = Arrays.binarySearch(mat[i], mid) + 1;
                midPos += pos;
            }

            if (midPos < medianPos) {
                min = mid + 1;
            } else {
                max = mid;
            }
        }

        return min;
    }

    public boolean isSudokuRowValid(String[] A, int rowStart, int rowEnd, int colStart, int colEnd) {
        for (int i = rowStart; i < rowEnd; i++) {

            HashSet<Character> s = new HashSet<Character>();

            for (int j = colStart; j < colEnd; j++) {

                char ch = A[i].charAt(j);
                if (ch != '.') {
                    if (s.contains(ch)) {
                        return false;
                    }

                    s.add(ch);
                }
            }
        }
        return true;
    }

    public boolean isSudokuColumnValid(String[] A, int rowStart, int rowEnd, int colStart, int colEnd) {

        for (int j = colStart; j < colEnd; j++) {

            HashSet<Character> s = new HashSet<Character>();

            for (int i = rowStart; i < rowEnd; i++) {
                char ch = A[i].charAt(j);

                if (ch != '.') {
                    if (s.contains(ch)) {
                        return false;
                    }

                    s.add(ch);
                }
            }
        }

        return true;
    }

    public boolean isSudokuMatrixValid(String[] A, int rowStart, int rowEnd, int colStart, int colEnd) {
        HashSet<Character> s = new HashSet<Character>();

        for (int i = rowStart; i < rowEnd; i++) {

            for (int j = colStart; j < colEnd; j++) {

                char ch = A[i].charAt(j);
                if (ch != '.') {
                    if (s.contains(ch)) {
                        return false;
                    }

                    s.add(ch);
                }
            }
        }

        return true;
    }

    // DO NOT MODIFY THE ARGUMENTS WITH "final" PREFIX. IT IS READ ONLY
    public int isValidSudoku(final String[] A) {

        if (!isSudokuRowValid(A, 0, 9, 0, 9))
            return 0;

        if (!isSudokuColumnValid(A, 0, 9, 0, 9))
            return 0;

        if (!isSudokuMatrixValid(A, 0, 3, 0, 3))
            return 0;

        if (!isSudokuMatrixValid(A, 0, 3, 3, 6))
            return 0;

        if (!isSudokuMatrixValid(A, 0, 3, 6, 9))
            return 0;

        int rowStart = 0, rowEnd = 3;

        while (rowEnd <= 9) {

            if (!isSudokuMatrixValid(A, rowStart, rowEnd, 0, 3))
                return 0;

            if (!isSudokuMatrixValid(A, rowStart, rowEnd, 3, 6))
                return 0;

            if (!isSudokuMatrixValid(A, rowStart, rowEnd, 6, 9))
                return 0;

            rowStart = rowStart + 3;
            rowEnd = rowEnd + 3;
        }

        return 1;
    }

    /* Find a peak element
    Given an array of integers A, find and return the peak element in it. An array element is peak if it is NOT smaller than its neighbors.

    For corner elements, we need to consider only one neighbor. We ensure that answer will be unique.

    NOTE: Users are expected to solve this in O(log(N)) time.



    Problem Constraints
    1 <= |A| <= 100000

    1 <= A[i] <= 109



    Input Format
    The only argument given is the integer array A.



    Output Format
    Return the peak element.



    Example Input
    Input 1:

    A = [1, 2, 3, 4, 5]
    Input 2:

    A = [5, 17, 100, 11]


    Example Output
    Output 1:

    5
    Output 2:

    100


    Example Explanation
    Explanation 1:

    5 is the peak.
    Explanation 2:

    100 is the peak.
    */
    public int findPeakElement(int[] A) {
        int n = A.length;
        if(n==1) return A[0];

        if(A[0] >= A[1]) return A[0];

        if(A[n-1] >= A[n-2]) return A[n-1];

        int l=1, r=n-2;
        
        while(l<=r)
        {
            int m = (l+r)/2;
            if(A[m] > A[m-1] && A[m] > A[m+1]) return A[m];

            if(A[m-1] > A[m])
            {
                r = m - 1;
            } else{
                l = m + 1;
            }
        }

        return -1;
    }

    /* Single Element in a Sorted Array
    Problem Description
    Given a sorted array of integers A where every element appears twice except for one element which appears once, find and return this single element that appears only once.

    NOTE: Users are expected to solve this in O(log(N)) time.



    Problem Constraints
    1 <= |A| <= 100000

    1 <= A[i] <= 10^9



    Input Format
    The only argument given is the integer array A.



    Output Format
    Return the single element that appears only once.



    Example Input
    Input 1:

    A = [1, 1, 7]
    Input 2:

    A = [2, 3, 3]


    Example Output
    Output 1:

    7
    Output 2:

    2


    Example Explanation
    Explanation 1:

    7 appears once
    Explanation 2:

    2 appears once
    */
    public int SingleElementInSortedArray(int[] A) {
        int n = A.length;

        if (A[n - 1] != A[n - 2])
            return A[n - 1];

        if (A[0] != A[1])
            return A[0];

        int l = 0, r = n - 1;
        while (l <= r) {
            int mid = (l + r) / 2;
            if (A[mid] != A[mid - 1] && A[mid] != A[mid + 1]) {
                return A[mid];
            }

            if (A[mid] == A[mid - 1]) {
                if (mid % 2 != 0) {
                    l = mid + 1;
                } else {
                    r = mid - 2;
                }

            } else {
                if (mid % 2 == 0) {
                    l = mid + 2;
                } else {
                    r = mid - 1;
                }
            }
        }

        return -1;
    }

    /* Square Root of Integer
    Given an integer A.

    Compute and return the square root of A.

    If A is not a perfect square, return floor(sqrt(A)).

    DO NOT USE SQRT FUNCTION FROM THE STANDARD LIBRARY.

    NOTE: Do not use the sqrt function from the standard library. Users are expected to solve this in O(log(A)) time.



    Problem Constraints
    0 <= A <= 1010



    Input Format
    The first and only argument given is the integer A.



    Output Format
    Return floor(sqrt(A))



    Example Input
    Input 1:

    11
    Input 2:

    9


    Example Output
    Output 1:

    3
    Output 2:

    3


    Example Explanation
    Explanation:

    When A = 11 , square root of A = 3.316. It is not a perfect square so we return the floor which is 3.
    When A = 9 which is a perfect square of 3, so we return 3.
    */
    public int sqrt(int A) {
        int l = 1, r = A;
        int ans = 0;
        while (l <= r) {
            int mid = (l + r) / 2;
            // mid * mid could overflow to avoid it, use trick
            if (mid == A / mid)
                return mid;

            if (mid > A / mid) {
                r = mid - 1;

            } else {
                ans = mid;
                l = mid + 1;
            }
        }

        return ans;
    }

    /* Rotated Sorted Array Search
    Given a sorted array of integers A of size N and an integer B.

    array A is rotated at some pivot unknown to you beforehand.

    (i.e., 0 1 2 4 5 6 7 might become 4 5 6 7 0 1 2 ).

    You are given a target value B to search. If found in the array, return its index otherwise, return -1.

    You may assume no duplicate exists in the array.

    NOTE: Users are expected to solve this in O(log(N)) time.



    Problem Constraints
    1 <= N <= 1000000

    1 <= A[i] <= 10^9

    all elements in A are distinct.



    Input Format
    The first argument given is the integer array A.

    The second argument given is the integer B.



    Output Format
    Return index of B in array A, otherwise return -1



    Example Input
    Input 1:

    A = [4, 5, 6, 7, 0, 1, 2, 3]
    B = 4 
    Input 2:

    A = [1]
    B = 1


    Example Output
    Output 1:

    0 
    Output 2:

    0


    Example Explanation
    Explanation 1:

    
    Target 4 is found at index 0 in A. 
    Explanation 2:

    
    Target 1 is found at index 0 in A.
    */
    
    public int SearchInSortedAndRotatedArray(final int[] A, int B) {
        if(B == A[0]) return 0;

        int n = A.length;

        if(B == A[n-1]) return n-1;

        int p = findPivotInSortedAndRotatedArray(A);

        if(p == -1)
        {
            return BinarySearchUtil(A, B, 0, n-1);
        }

        if(B > A[0])
        {
            // lies in first half
            return BinarySearchUtil(A, B, 0, p-1);
        } 
        
        // lies in second half
        return BinarySearchUtil(A, B, p, n-1);
    }

    /* Find Pivot in sorted and rotated array*/
    public int findPivotInSortedAndRotatedArray(int[] A)
    {
        // pivot is the index of smallest element in sorted and rotated array A
        int l=0, r=A.length-1;

        while(l<=r)
        {
            int mid = (l+r)/2;

            if(A[mid] > A[0])
            {
                // mid is in first half and pivot cannot be in first half
                l = mid + 1;
            } else {
                // mid is in second half
                if(A[mid] < A[mid-1]) return mid;

                r = mid - 1;                
            }
        }

        return -1;
    }
}
